### PR TITLE
*: set write and time details in RPC responses 

### PR DESCRIPTION
**Pull request by sticnarf**
### What is changed and how it works?

Issue Number: Ref https://github.com/tikv/tikv/issues/12362

Depends on https://github.com/pingcap/kvproto/pull/931

What's Changed:

```commit-message
kvproto#931 adds a few more details about the time used
by an RPC request. We are filling these additional information
to the responses in this commit.
```

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by sticnarf** [comment]
@you06 @cfzjywxk I think this PR is reviewable but some fields are added in https://github.com/pingcap/kvproto/pull/931. And you can also refer to https://github.com/tikv/client-go/pull/536 and https://github.com/pingcap/tidb/pull/35814 for using them in the TiDB slow log.

File: src/server/service/kv.rs [hunk 1]
```diff
@@ -369,7 +400,7 @@ impl<T: RaftStoreRouter<E::Local> + 'static, E: Engine, L: LockManager, F: KvFor
 
     fn coprocessor(&mut self, ctx: RpcContext<'_>, req: Request, sink: UnarySink<Response>) {
         forward_unary!(self.proxy, coprocessor, ctx, req, sink);
-        let begin_instant = Instant::now_coarse();
+        let begin_instant = Instant::now();
```
Comment (cfzjywxk) [inline src/server/service/kv.rs:403]: Why change this here? It seems the usages are not consistent for example the snapshot duration calculation.

Comment (sticnarf) [reply src/server/service/kv.rs:403]: now_coarse can only provides 1~10ms precision (depending on the kernel config), which is sometimes not accurate enough for us.

**Comment by sticnarf** [comment]
The total rpc time is also returned in the response. So, we can know the exact time used in the RPC layer (including the network latency).

image

File: src/server/service/batch.rs [hunk 2]
```diff
@@ -35,7 +35,7 @@ pub struct ReqBatcher {
 
 impl ReqBatcher {
     pub fn new(batch_size: usize) -> ReqBatcher {
-        let begin_instant = Instant::now_coarse();
+        let begin_instant = Instant::now();
```
Comment (siddontang) [inline src/server/service/batch.rs:38]: may using now here affect the performance in the hot path?

Comment (sticnarf) [reply src/server/service/batch.rs:38]: `now` takes about 20ns while `now_coarse` takes about 4ns.

As our fastest RPC takes tens of microseconds, such increase is ignorable.

**Comment by you06** [comment]
> Maybe we could add a new kv time graph to show these different layers somewhere together:

Maybe it's good to display such metrics in performance overview dashboard.

File: components/tracker/src/lib.rs [hunk 3]
```diff
@@ -34,9 +34,38 @@ impl Tracker {
     pub fn write_scan_detail(&self, detail_v2: &mut pb::ScanDetailV2) {
         detail_v2.set_rocksdb_block_read_byte(self.metrics.block_read_byte);
         detail_v2.set_rocksdb_block_read_count(self.metrics.block_read_count);
+        detail_v2.set_rocksdb_block_read_nanos(self.metrics.block_read_nanos);
```
Comment (you06) [inline components/tracker/src/lib.rs:37]: Also need to update `metrics.block_read_nanos` when reporting metrics to the tracker?

File: components/tracker/src/lib.rs [hunk 4]
```diff
@@ -34,9 +34,38 @@ impl Tracker {
...
         detail_v2.set_rocksdb_block_cache_hit_count(self.metrics.block_cache_hit_count);
         detail_v2.set_rocksdb_key_skipped_count(self.metrics.internal_key_skipped_count);
         detail_v2.set_rocksdb_delete_skipped_count(self.metrics.deleted_key_skipped_count);
+        detail_v2.set_get_snapshot_nanos(self.metrics.get_snapshot_nanos);
+    }
+
+    pub fn write_write_detail(&self, detail: &mut pb::WriteDetail) {
+        detail.set_store_batch_wait_nanos(self.metrics.wf_batch_wait_nanos);
```
Comment (cfzjywxk) [inline components/tracker/src/lib.rs:45]: Do we need to link the performance map in the comments to explain the relationships?

**Review by cfzjywxk — APPROVED** [review]

**Review by you06 — APPROVED** [review]

File: components/tracker/src/lib.rs (hunk 4 above)
Comment (sticnarf) [reply components/tracker/src/lib.rs:45]: Let me the change some bits of the performance to match the metrics first.

**Pull request merged**